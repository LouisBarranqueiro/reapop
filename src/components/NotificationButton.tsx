import React, {useState} from 'react'
import {Notification, NotificationButton as ButtonType} from '../reducers/notifications/types'
import {Theme} from '../themes/types'
import {classnames} from '../constants'

type Props = {
    button: ButtonType
    position: number
    notification: Notification
    theme?: Theme
}

export const NotificationButton = (props: Props) => {
    const {notification, theme, position, button} = props
    const [isHovered, setIsHovered] = useState(false)
    const [isActive, setIsActive] = useState(false)
    const state = {isHovered, isActive}
    const buttonStyles = theme ? theme.notificationButton(notification, position, state) : {}
    const buttonTextStyles = theme ? theme.notificationButtonText(notification, position, state) : {}

    return (
        <button
            data-testid="button"
            style={buttonStyles}
            className={classnames.notificationButton}
            onMouseLeave={() => setIsHovered(false)}
            onMouseEnter={() => setIsHovered(true)}
            onMouseDown={() => setIsActive(true)}
            onMouseUp={() => setIsActive(false)}
        >
            <span className={classnames.notificationButtonText} style={buttonTextStyles}>
                {button.primary ? <b>{button.name}</b> : button.name}
            </span>
        </button>
    )
}

export default NotificationButton
