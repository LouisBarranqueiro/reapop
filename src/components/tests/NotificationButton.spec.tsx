import React from 'react'
import {act, render, fireEvent} from '@testing-library/react'

import atalhoTheme from '../../themes/atalho'
import {Notification, POSITIONS, STATUSES} from '../..'
import {NotificationButton} from '../NotificationButton'

describe('<NotificationButton/>', () => {
    const notification: Notification = {
        id: '1',
        position: POSITIONS.topLeft,
        status: STATUSES.none,
        buttons: [],
    }
    beforeEach(() => {
        jest.spyOn(atalhoTheme, 'notificationButton').mockReturnValue({color: 'red'})
        jest.spyOn(atalhoTheme, 'notificationButtonText').mockReturnValue({color: 'red'})
    })

    afterEach(() => {
        jest.clearAllMocks()
    })

    afterAll(() => {
        jest.resetAllMocks()
    })

    it.each([
        ['without theme', undefined],
        ['with theme', atalhoTheme],
    ])('should render button %s', (_, theme) => {
        const {container} = render(
            <NotificationButton button={{name: 'yes'}} position={2} notification={notification} theme={theme} />
        )
        expect((atalhoTheme.notificationButton as jest.Mock).mock.calls).toMatchSnapshot('notificationButton.calls')
        expect((atalhoTheme.notificationButtonText as jest.Mock).mock.calls).toMatchSnapshot(
            'notificationButtonText.calls'
        )
        expect(container.innerHTML).toMatchSnapshot()
    })

    it.each([
        ['secondary', false],
        ['primary', true],
    ])('should render a %s button ', (_, primary) => {
        const {container} = render(
            <NotificationButton button={{name: 'yes', primary}} position={1} notification={notification} />
        )
        expect(container.innerHTML).toMatchSnapshot()
    })

    it('should update button when hovering it and clicking it', () => {
        const {container, getByTestId} = render(
            <NotificationButton button={{name: 'yes'}} position={1} notification={notification} theme={atalhoTheme} />
        )
        const button = getByTestId('button')
        ;['mouseEnter', 'mouseDown', 'mouseUp', 'mouseLeave'].map((eventType) => {
            jest.clearAllMocks()
            act(() => {
                fireEvent[eventType as 'mouseEnter' | 'mouseDown' | 'mouseUp' | 'mouseLeave'](button)
            })
            expect((atalhoTheme.notificationButton as jest.Mock).mock.calls).toMatchSnapshot('notificationButton.calls')
            expect((atalhoTheme.notificationButtonText as jest.Mock).mock.calls).toMatchSnapshot(
                'notificationButtonText.calls'
            )
            expect(container.innerHTML).toMatchSnapshot(eventType)
        })

        jest.resetAllMocks()
    })
})
