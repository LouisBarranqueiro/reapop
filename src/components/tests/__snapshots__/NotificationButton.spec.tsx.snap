// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationButton/> should render a primary button  1`] = `"<button data-testid=\\"button\\" class=\\"reapop__notification-button\\"><span class=\\"reapop__notification-button-text\\"><b>yes</b></span></button>"`;

exports[`<NotificationButton/> should render a secondary button  1`] = `"<button data-testid=\\"button\\" class=\\"reapop__notification-button\\"><span class=\\"reapop__notification-button-text\\">yes</span></button>"`;

exports[`<NotificationButton/> should render button with theme 1`] = `"<button data-testid=\\"button\\" style=\\"color: red;\\" class=\\"reapop__notification-button\\"><span class=\\"reapop__notification-button-text\\" style=\\"color: red;\\">yes</span></button>"`;

exports[`<NotificationButton/> should render button with theme: notificationButton.calls 1`] = `
Array [
  Array [
    Object {
      "buttons": Array [],
      "id": "1",
      "position": "top-left",
      "status": "none",
    },
    2,
    Object {
      "isActive": false,
      "isHovered": false,
    },
  ],
]
`;

exports[`<NotificationButton/> should render button with theme: notificationButtonText.calls 1`] = `
Array [
  Array [
    Object {
      "buttons": Array [],
      "id": "1",
      "position": "top-left",
      "status": "none",
    },
    2,
    Object {
      "isActive": false,
      "isHovered": false,
    },
  ],
]
`;

exports[`<NotificationButton/> should render button without theme 1`] = `"<button data-testid=\\"button\\" class=\\"reapop__notification-button\\"><span class=\\"reapop__notification-button-text\\">yes</span></button>"`;

exports[`<NotificationButton/> should render button without theme: notificationButton.calls 1`] = `Array []`;

exports[`<NotificationButton/> should render button without theme: notificationButtonText.calls 1`] = `Array []`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: mouseDown 1`] = `"<button data-testid=\\"button\\" style=\\"color: red;\\" class=\\"reapop__notification-button\\"><span class=\\"reapop__notification-button-text\\" style=\\"color: red;\\">yes</span></button>"`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: mouseEnter 1`] = `"<button data-testid=\\"button\\" style=\\"color: red;\\" class=\\"reapop__notification-button\\"><span class=\\"reapop__notification-button-text\\" style=\\"color: red;\\">yes</span></button>"`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: mouseLeave 1`] = `"<button data-testid=\\"button\\" style=\\"color: red;\\" class=\\"reapop__notification-button\\"><span class=\\"reapop__notification-button-text\\" style=\\"color: red;\\">yes</span></button>"`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: mouseUp 1`] = `"<button data-testid=\\"button\\" style=\\"color: red;\\" class=\\"reapop__notification-button\\"><span class=\\"reapop__notification-button-text\\" style=\\"color: red;\\">yes</span></button>"`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: notificationButton.calls 1`] = `
Array [
  Array [
    Object {
      "buttons": Array [],
      "id": "1",
      "position": "top-left",
      "status": "none",
    },
    1,
    Object {
      "isActive": false,
      "isHovered": true,
    },
  ],
]
`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: notificationButton.calls 2`] = `
Array [
  Array [
    Object {
      "buttons": Array [],
      "id": "1",
      "position": "top-left",
      "status": "none",
    },
    1,
    Object {
      "isActive": true,
      "isHovered": true,
    },
  ],
]
`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: notificationButton.calls 3`] = `
Array [
  Array [
    Object {
      "buttons": Array [],
      "id": "1",
      "position": "top-left",
      "status": "none",
    },
    1,
    Object {
      "isActive": false,
      "isHovered": true,
    },
  ],
]
`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: notificationButton.calls 4`] = `
Array [
  Array [
    Object {
      "buttons": Array [],
      "id": "1",
      "position": "top-left",
      "status": "none",
    },
    1,
    Object {
      "isActive": false,
      "isHovered": false,
    },
  ],
]
`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: notificationButtonText.calls 1`] = `
Array [
  Array [
    Object {
      "buttons": Array [],
      "id": "1",
      "position": "top-left",
      "status": "none",
    },
    1,
    Object {
      "isActive": false,
      "isHovered": true,
    },
  ],
]
`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: notificationButtonText.calls 2`] = `
Array [
  Array [
    Object {
      "buttons": Array [],
      "id": "1",
      "position": "top-left",
      "status": "none",
    },
    1,
    Object {
      "isActive": true,
      "isHovered": true,
    },
  ],
]
`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: notificationButtonText.calls 3`] = `
Array [
  Array [
    Object {
      "buttons": Array [],
      "id": "1",
      "position": "top-left",
      "status": "none",
    },
    1,
    Object {
      "isActive": false,
      "isHovered": true,
    },
  ],
]
`;

exports[`<NotificationButton/> should update button when hovering it and clicking it: notificationButtonText.calls 4`] = `
Array [
  Array [
    Object {
      "buttons": Array [],
      "id": "1",
      "position": "top-left",
      "status": "none",
    },
    1,
    Object {
      "isActive": false,
      "isHovered": false,
    },
  ],
]
`;
